bpf/lib/fib.h

```diff
@@ -8,134 +8,181 @@
+ * or L3->L2 device.
+ */
+static __always_inline int
+fib_redirect(struct __ctx_buff *ctx, const bool needs_l2_check,
+	     struct bpf_fib_lookup_padded *fib_params, __s8 *fib_err, int *oif)
 {
+	struct bpf_redir_neigh nh_params;
+	struct bpf_redir_neigh *nh = NULL;
 	bool no_neigh = is_defined(ENABLE_SKIP_FIB);
 	int ret;
-	struct bpf_redir_neigh *nh = NULL;
-# ifndef ENABLE_SKIP_FIB
-	struct bpf_redir_neigh nh_params;
-	struct bpf_fib_lookup fib_params = {
-		.family		= AF_INET6,
-		.ifindex	= ctx->ingress_ifindex,
-	};
-
-	ipv6_addr_copy((union v6addr *)&fib_params.ipv6_src,
-		       (union v6addr *)&ip6->saddr);
-	ipv6_addr_copy((union v6addr *)&fib_params.ipv6_dst,
-		       (union v6addr *)&ip6->daddr);
 
-	ret = fib_lookup(ctx, &fib_params, sizeof(fib_params),
+#ifndef ENABLE_SKIP_FIB
+	ret = fib_lookup(ctx, &fib_params->l, sizeof(fib_params->l),
 			 BPF_FIB_LOOKUP_DIRECT);
-	switch (ret) {
-	case BPF_FIB_LKUP_RET_SUCCESS:
-		break;
-	case BPF_FIB_LKUP_RET_NO_NEIGH:
-		nh_params.nh_family = fib_params.family;
-		__bpf_memcpy_builtin(&nh_params.ipv6_nh, &fib_params.ipv6_dst,
-				     sizeof(nh_params.ipv6_nh));
-		no_neigh = true;
-		nh = &nh_params;
-		break;
-	default:
-		return CTX_ACT_DROP;
+	if (ret != BPF_FIB_LKUP_RET_SUCCESS) {
+		*fib_err = (__s8)ret;
+		if (likely(ret == BPF_FIB_LKUP_RET_NO_NEIGH)) {
+			nh_params.nh_family = fib_params->l.family;
+			__bpf_memcpy_builtin(&nh_params.ipv6_nh,
+					     &fib_params->l.ipv6_dst,
+					     sizeof(nh_params.ipv6_nh));
+			nh = &nh_params;
+			no_neigh = true;
+			/* For kernels without d1c362e1dd68 ("bpf: Always
+			 * return target ifindex in bpf_fib_lookup") we
+			 * fall back to use the caller-provided oif when
+			 * necessary.
+			 */
+			if (!is_defined(HAVE_FIB_IFINDEX) && *oif > 0)
+				goto skip_oif;
+		} else {
+			return DROP_NO_FIB;
+		}
 	}
-
-	*oif = fib_params.ifindex;
-# else
+	*oif = fib_params->l.ifindex;
+skip_oif:
+#else
 	*oif = DIRECT_ROUTING_DEV_IFINDEX;
-# endif /* ENABLE_SKIP_FIB */
+	nh_params.nh_family = fib_params->l.family;
+#endif /* ENABLE_SKIP_FIB */
+	if (needs_l2_check) {
+		bool l2_hdr_required = true;
 
-	ret = ipv6_l3(ctx, l3_off, NULL, NULL, METRIC_EGRESS);
-	if (unlikely(ret != CTX_ACT_OK))
-		return ret;
+		ret = maybe_add_l2_hdr(ctx, *oif, &l2_hdr_required);
+		if (ret != 0)
+			return ret;
+		if (!l2_hdr_required)
+			goto out_send;
+	}
 	if (no_neigh) {
-		if (nh)
-			return redirect_neigh(*oif, nh, sizeof(*nh), 0);
-		else
-			return redirect_neigh(*oif, NULL, 0, 0);
+		/* If we are able to resolve neighbors on demand, always
+		 * prefer that over the BPF neighbor map since the latter
+		 * might be less accurate in some asymmetric corner cases.
+		 */
+		if (neigh_resolver_available()) {
+			if (nh)
+				return redirect_neigh(*oif, &nh_params,
+						      sizeof(nh_params), 0);
+			else
+				return redirect_neigh(*oif, NULL, 0, 0);
+		} else {
+			union macaddr *dmac, smac = NATIVE_DEV_MAC_BY_IFINDEX(*oif);
+
```

Comment:
The macros are generated here based on all the phys devices that were identified (`option.Config.GetDevices()`): https://github.com/cilium/cilium/blob/e337ebd684491f658724c832ef3a6279c10a855d/pkg/datapath/linux/config/config.go#L715 . If something is missing in NATIVE_DEV_MAC_BY_IFINDEX, then it will fallback to default `union macaddr __mac = {.addr = {0x0, 0x0, 0x0, 0x0, 0x0, 0x0}};` . If something is missing in IS_L3_DEV it will fallback to false.

Reply:
So code paths that previously had *no* fallback for missing L2 resolution now have a fallback, which requires proper `--devices` setup. Thanks!  :+1:

Reply:
> So code paths that previously had _no_ fallback for missing L2 resolution now have a fallback, which requires proper `--devices` setup. Thanks! 👍

Yeah, the core is that if redirect_neigh is available, we will use this to driver L2 resolution in all the places which will fix issues like https://github.com/cilium/cilium/issues/22782 . But for cases on older kernels where it is not available (or for XDP layer), the only other option is to drive L2 based on the neighbor map. The next step to complete this is on the control plane side to push down all service backends into the neighbor management of the agent. We only do this for all nodes in the cluster if they're in the same L2 via managed neighbors (https://github.com/cilium/cilium/pull/17713), but we also need it for service backends if they're not part of the cluster itself.

Reply:
> > So code paths that previously had _no_ fallback for missing L2 resolution now have a fallback, which requires proper `--devices` setup. Thanks! +1
> 
> Yeah setup or auto-detection; the core is that if redirect_neigh is available, we will use this to drive L2 resolution in all the places which will fix issues like #22782 . 

Would you want to backport to v1.13 then?

Reply:
It would be good, yes. I'd say we let the series sit for a few weeks in master and then I can do the backport myself for 1.13.